### ISSUE TITLE
How we can retrain KNN for 1000 of faces and keep the performance-optimized.

### ISSUE THREAD

**Issue by Anki7s** [ageitgey/face_recognition#949]
* face_recognition version:1.2.3
* Python version: 3.6.7
* Operating System: windows-10

### Description
@ageitgey 
How we can retrain KNN for 1000 of faces and keep the performance-optimized for cpu. 

 

### What I Did

I tried to retrain the KNN model on  500 images but when I add new face images the training process start from the beginning. 

Is there any way to start the training process only for new images (faces) instead of all.

**Comment by alessiosavi** [comment]
KNN is not very suitable for face recognition. If you change the `jitter` parameter, it will confuse the network.

Have a try with my little project that wrap `face_recognition` in order to extract the face points, and than use a neural network (MLP) in order to classify the given face.

**Comment by Anki7s** [comment]
> 

> 
> 
> KNN is not very suitable for face recognition. If you change the `jitter` parameter, it will confuse the network.
> 
> Have a try with my little project that wrap `face_recognition` in order to extract the face points, and than use a neural network (MLP) in order to classify the given face.

Thanks @alessiosavi  for sharing your knowledge. will surly check it.

**Comment by alessiosavi** [comment]
Due to the nature of the KNN, it's quite impossible to find the K parameter that is suitable for different faces.

If your dataset contains only a few dozen of different people, maybe you can find it with a exhaustive hyperparameter tuning.

You can rely on the `Multi Layer Perceptron` architecture in order to create a neural network capable of recognize different faces with higher accuracy.

I've made a little project that use this library (`face_recognition`), linked with a Neural Network. It's awesome see that, starting from 4/5 photos of a person, the network is able to recognize the person using different photos.

You can have a look at my repository:

https://github.com/alessiosavi/PyRecognizer/

**Comment by sandip-narwade** [comment]
> KNN is not very suitable for face recognition. If you change the `jitter` parameter, it will confuse the network.
> 
> Have a try with my little project that wrap `face_recognition` in order to extract the face points, and than use a neural network (MLP) in order to classify the given face.

How to use MPL

**Comment by alessiosavi** [comment]
You can use my `PyRecognizer` in order to have a plug-and-play face-recognition tool, without deal with coding. The repository have a video for understand how to train the neural network and how to predict an image (the model provided is tuned for some celebrities). Or you can view the code that i have used here: https://github.com/alessiosavi/PyRecognizer/blob/master/datastructure/Classifier.py#L332

Link to the project: https://github.com/alessiosavi/PyRecognizer

**Comment by sandip-narwade** [comment]
@alessiosavi , While predicting it is taking almost 58 seconds. It too late response,  @ageitgey face_recognition took approximate 2.5 seconds using Flask REST API. 

Can I make it faster?

**Comment by alessiosavi** [comment]
Please open an issue on the repo. Have you a GPU? The problems seems related to the `encoding_faces` phase. You can see that the most of the time is spent
